File: src/validate/no_global_duplicates.rs
use std::collections::HashMap;

use crate::models::bundle::Bundle;

pub fn run(bundles: &Vec<Bundle>) -> Result<(), String> {
    // Checking for duplicated bundle names
    {
        let mut seen: HashMap<String, String> = HashMap::new(); // name -> base_url
        for b in bundles {
            if seen.contains_key(&b.name) {
                return Err(format!(
                    "ERROR: {}.{} Duplicated-Bundle-Name url={} error={}",
                    file!(),
                    line!(),
                    b.base_url,
                    b.name,
                ));
            }
            seen.insert(b.name.clone(), b.base_url.clone());
        }
    }

    // Checking for duplicated source names in the UI
    {
        let mut seen: HashMap<String, String> = HashMap::new(); // full_title -> bundle name
        for b in bundles {
            if let Some(existing_name) = seen.get(&b.ui.source.full_title) {
                if existing_name != &b.name {
                    return Err(format!(
                        "ERROR: {}.{} Duplicated-UI-Source-Name url={} error={}",
                        file!(),
                        line!(),
                        b.base_url,
                        b.ui.source.full_title
                    ));
                }
            } else {
                seen.insert(b.ui.source.full_title.clone(), b.name.clone());
            }
        }
    }

    // Checking for duplicated base_urls (each must be unique)
    {
        let mut seen: HashMap<String, usize> = HashMap::new();
        for b in bundles {
            *seen.entry(b.base_url.clone()).or_insert(0) += 1;
        }
        for (url, count) in &seen {
            if *count > 1 {
                return Err(format!(
                    "ERROR: {}.{} Duplicated-Base-URL count={} url={}",
                    file!(),
                    line!(),
                    count,
                    url
                ));
            }
        }
    }

    Ok(())
}

File: bundle-deployer/src/validation/no_global_duplicates.ts
// Validation: Check for duplicates across ALL bundles

import type { Bundle } from "../types/bundle.ts";

export function run(bundles: Bundle[]): void {
  // Check for duplicated bundle names
  {
    const tokens = new Set<string>();
    for (const bundle of bundles) {
      if (tokens.has(bundle.name)) {
        throw new Error(
          `Duplicated-Bundle-Name url=${bundle.base_url} error=${bundle.name}`
        );
      }
      tokens.add(bundle.name);
    }
  }
  
  // Check for duplicated source names in the UI
  {
    const tokens = new Set<string>();
    for (const bundle of bundles) {
      if (tokens.has(bundle.ui.source.full_title)) {
        throw new Error(
          `Duplicated-UI-Source-Name url=${bundle.base_url} error=${bundle.ui.source.full_title}`
        );
      }
      tokens.add(bundle.ui.source.full_title);
    }
  }
  
  // Check for duplicated names across all bundles
  const tokens = new Map<string, number>();
  
  for (const bundle of bundles) {
    tokens.set(bundle.name, (tokens.get(bundle.name) || 0) + 1);
    
    for (const table of bundle.tables) {
      tokens.set(table.name, (tokens.get(table.name) || 0) + 1);
    }
    
    tokens.set(
      bundle.ui.source.full_title,
      (tokens.get(bundle.ui.source.full_title) || 0) + 1
    );
    
    tokens.set(bundle.base_url, (tokens.get(bundle.base_url) || 0) + 1);
  }
  
  // Check for duplicates
  for (const [name, count] of tokens.entries()) {
    if (count > 1) {
      throw new Error(`Duplicated-Name count=${count} table=${name}`);
    }
  }
}